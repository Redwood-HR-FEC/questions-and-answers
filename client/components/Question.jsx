import React from 'react';
import styled from 'styled-components';

const QuestionWrapper = styled.div`
  margin-bottom: 10px!important;
`;

const ATag = styled.a`
  color:rgb(0, 102, 192);
  text-decoration-line:none;
  text-decoration-style:solid;

  &:hover {
    color:rgb(196, 85, 0);
    text-decoration-line:underline;
    text-decoration-style:solid;
  }
`;

const Question = ({ question, url }) => {
  return (
    <QuestionWrapper>
      <ATag href={url}>
        { question }
        ?
      </ATag>
    </QuestionWrapper>
  );
};

export default Question;
