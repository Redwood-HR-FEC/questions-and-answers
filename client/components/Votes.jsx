import React from 'react';
import styled from 'styled-components';
import Question from './Question.jsx';
import Answer from './Answer.jsx';

const ListUp = styled.li`
  background-image: url();
  list-style-type: none;
  border-image-repeat:stretch;
  box-sizing:border-box;
  align-items:flex-start;
`;

const ImageUp = styled.img`
  &:hover{
    content: url(https://m.media-amazon.com/images/G/01/x-locale/communities/discussion_boards/highlighted_up_arrow._CB442978120_.png);
  }
`;

const ImageDown = styled.img`
  &:hover{
    content: url(https://m.media-amazon.com/images/G/01/x-locale/communities/discussion_boards/highlighted_down_arrow._CB442978121_.png);
  }
`;
const ListDown = styled.li`
  background-image: url();
  list-style-type: none;
  border-image-repeat:stretch;
  box-sizing:border-b ox;
  align-items:flex-start;
  margin-top: 5px;

  &:hover {
    background-image: url();
  }
`;
const QuestionAnswerWraper = styled.div`
  box-sizing:border-box;
  position:relative;
  padding-left:7.34375px;
  margin-top: 20px;
`;
const Vote = styled.ul`
  color:rgb(17, 17, 17);
  display:block;
  float:left;
  border-right: 1px solid #ddd;
  width:65px;
  margin-left:-35px;
  line-height: 19px;
  text-align:center;
  padding-right: 15px;
`;

const WrapperBox = styled.div`
  box-sizing:border-box;
  color:rgb(17, 17, 17);

`;

const ShowVotes = styled.li`
  box-sizing:border-box;
  color:rgb(0, 0, 0);
  display:inline;

`;
const QuestionWrapper = styled.div`
  margin-right: 1.5rem;
  font-weight:700;
`;

const AnswersWrapper = styled.div`
  margin-right: 2rem;
  box-sizing:border-box;
  font-weight:700;
`;

const VotesQuestionsAnswers = styled.div`
  display: flex;
`;

const Wrap = styled.div`
  display: flex;
`;

class Votes extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      up: 0,
      down: 0,
    };

    this.handleUpvotes = this.handleUpvotes.bind(this);
    this.handleDownvotes = this.handleDownvotes.bind(this);
  }

  handleUpvotes(id, votes, productId) {
    // console.log(id);
    const { changeVote } = this.props;
    changeVote(votes + 1, id, productId);
  }

  handleDownvotes(id, votes, productId) {
    const { changeVote } = this.props;
    changeVote(votes - 1, id, productId);
  }


  render() {
    const { data } = this.props;

    return (
      <WrapperBox>

        {data.map((question) => {
          if (question.answers.length > 0) {
            return (
              <VotesQuestionsAnswers key={question.product_id + question.votes}>
                <Vote>
                  <ListUp onClick={() => this.handleUpvotes(question._id, question.votes, question.product_id)}>
                    <ImageUp src="https://m.media-amazon.com/images/G/01/x-locale/communities/discussion_boards/neutral_up_arrow._CB442978120_.png" alt="Up" />
                  </ListUp>
                  <ShowVotes>
                    { question.votes }
                    <br />
                    <span>votes</span>
                  </ShowVotes>
                  <ListDown onClick={()=> this.handleDownvotes(question._id, question.votes, question.product_id)}>
                    <ImageDown src="https://m.media-amazon.com/images/G/01/x-locale/communities/discussion_boards/neutral_down_arrow._CB442978120_.png" alt="Down" />
                  </ListDown>
                </Vote>
                <QuestionAnswerWraper>
                  <Wrap>
                    <QuestionWrapper>Question:</QuestionWrapper>
                    <Question question={question.questions} url={question.question_url} />
                  </Wrap>
                  <Wrap>
                    <AnswersWrapper>Answer:</AnswersWrapper>

                    <Answer answer={question.answers} />
                  </Wrap>
                </QuestionAnswerWraper>
              </VotesQuestionsAnswers>
            );
          }
        })}
      </WrapperBox>
    );
  }
}

export default Votes;
