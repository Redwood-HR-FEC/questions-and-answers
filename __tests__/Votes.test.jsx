import React from 'react';
import { shallow } from 'enzyme';
import Votes from '../client/components/Votes';
import Question from '../client/components/Question';

const sampleData = [
  {
    "_id" : "5eb0aa58fdb2ea2b7daec7f3",
    "answers" : [
      {
        "username" : "Ahmad.Bruen",
        "answer" : "Ergonomic green workforce Row 24/365 initiative Divide JBOD Money Market Account circuit Ergonomic Wooden Shoes Ergonomic backing up Handcrafted Radial wireless hub optical Up-sized Aruba",
        "date_posted" : "2019-12-07T17:14:54.404Z",
      },
      {
        "username" : "Vivien75",
        "answer" : "Automated deposit silver Strategist Division Berkshire Business-focused Montana deposit Bahamas Rustic Plastic relationships integrated Oklahoma collaborative Chips reboot primary Lebanese Pound",
        "date_posted" : "2019-10-24T15:58:43.321Z",
      }
    ],
    "product_id" : "068",
    "questions" : "tan Center Kansas Comoro Franc Developer sky blue Wooden open-source Designer Lead",
    "question_url" : "http://larissa.name",
    "votes" : 26,
    "__v" : 0
  },
];

describe('Unit test for Votes component', () => {
  test('should render votes, answers and questions', () => {
    const wrapper = shallow(
      <Votes
        data={sampleData}
        changeVote={() => {}}
      />
    );
    expect(wrapper).toExist();
  });

  it ('accepts ussampleDataer props', () => {
    const wrapper = mount(
      <Votes
        data={sampleData}
        changeVote={() => {}}
      />,
    );
    expect(wrapper.props().data).toEqual(sampleData);
  });

  it("renders correctly", () => {
    const wrapper = shallow(
      <Votes
        data={sampleData}
        changeVote={() => {}}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Mouse events', () => {
  test('mouseclick methods to be instance of Votes', () => {
    const wrapper = shallow(
      <Votes
        data={sampleData}
        changeVote={() => {}}
      />,
    );
    wrapper.instance().handleUpvotes();
    wrapper.instance().handleDownvotes();
  });

  test('should update votes  when click li tag', () => {
    const mockChangeVote = jest.fn();
    const wrapper = mount(
      <Votes
        data={sampleData}
        changeVote={mockChangeVote}
      />,
    );
    const liTagUp = wrapper.find('img').first();
    liTagUp.simulate('click');
    expect(mockChangeVote).toHaveBeenCalled();

    const liTagDown = wrapper.find('img').last();
    liTagDown.simulate('click');
    expect(mockChangeVote).toHaveBeenCalled();
  });

  // test('should update votes -1 when click vote down li tag', () => {
  //   const mockChangeVote = jest.fn();
  //   const wrapper = mount(
  //     <Votes
  //       data={sampleData}
  //       changeVote={mockChangeVote}
  //     />,
  //   );
  //   const liTag = wrapper.find('img').last();
  //   liTag.simulate('click');
  //   expect(mockChangeVote).toHaveBeenCalled();
  // });
});

describe('Unit test for Question component', () => {
  const sampleQuestion = 'tan Center Kansas Comoro Franc Developer sky blue Wooden open-source Designer Lead';
  const sampleURL = 'http://larissa.name';
  test('should render 1 question', () => {
    const wrapper = shallow(
      <Question
        question={sampleQuestion}
        url={sampleURL}
      />
    );
    expect(wrapper).toHaveLength(1);
  });
});
