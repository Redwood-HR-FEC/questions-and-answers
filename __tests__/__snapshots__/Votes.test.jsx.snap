// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for Votes component renders correctly 1`] = `
<Votes__WrapperBox>
  <Votes__VotesQuestionsAnswers
    key="06826"
  >
    <Votes__Vote>
      <Votes__ListUp
        className="upVote"
        onClick={[Function]}
      >
        <Votes__ImageUp
          alt="Up"
          src="https://m.media-amazon.com/images/G/01/x-locale/communities/discussion_boards/neutral_up_arrow._CB442978120_.png"
        />
      </Votes__ListUp>
      <Votes__ShowVotes>
        26
        <br />
        <span>
          votes
        </span>
      </Votes__ShowVotes>
      <Votes__ListDown
        onClick={[Function]}
      >
        <Votes__ImageDown
          alt="Down"
          src="https://m.media-amazon.com/images/G/01/x-locale/communities/discussion_boards/neutral_down_arrow._CB442978120_.png"
        />
      </Votes__ListDown>
    </Votes__Vote>
    <Votes__QuestionAnswerWraper>
      <Votes__Wrap>
        <Votes__QuestionWrapper>
          Question:
        </Votes__QuestionWrapper>
        <Question
          question="tan Center Kansas Comoro Franc Developer sky blue Wooden open-source Designer Lead"
          url="http://larissa.name"
        />
      </Votes__Wrap>
      <Votes__Wrap>
        <Votes__AnswersWrapper>
          Answer:
        </Votes__AnswersWrapper>
        <Answer
          answer={
            Array [
              Object {
                "answer": "Ergonomic green workforce Row 24/365 initiative Divide JBOD Money Market Account circuit Ergonomic Wooden Shoes Ergonomic backing up Handcrafted Radial wireless hub optical Up-sized Aruba",
                "date_posted": "2019-12-07T17:14:54.404Z",
                "username": "Ahmad.Bruen",
              },
              Object {
                "answer": "Automated deposit silver Strategist Division Berkshire Business-focused Montana deposit Bahamas Rustic Plastic relationships integrated Oklahoma collaborative Chips reboot primary Lebanese Pound",
                "date_posted": "2019-10-24T15:58:43.321Z",
                "username": "Vivien75",
              },
            ]
          }
        />
      </Votes__Wrap>
    </Votes__QuestionAnswerWraper>
  </Votes__VotesQuestionsAnswers>
</Votes__WrapperBox>
`;
